### PR TITLE
[Image] Support passing AMIs for different regions

### LINKED ISSUE #1345: Support passing multiple `image_id`, one per region
From user, whose app requires a specific Ubuntu 18.04 version, and does not support the latest 20.04 we use:

> An alternative could be to allow specifying image_id as a mapping from region to AMI so that at least the user won’t be limited to specifying a fixed region. This would be more flexible (esp if user wants to use custom AMIs) but a bit more work on the user’s side.

Previously, @pounde also requested this feature.

-----

Proposal:
```
resources:
  image_id:
    us-east-1: <find your us-east-1 AMI>
    us-west-1: <find your us-west-1 AMI>
```
where SkyPilot will restrict itself to provision in these regions only.

### PR DESCRIPTION
**Pull request by Michaelvll**
Closes #1345 

Main changes:
1. Support a dict for `image_id`
2. Added 3 smoke tests for the image_id dict.

Notes:
1. The failover will only go through the regions listed in the `image_id`.
3. If a region or zone is specified and not in the `image_ids`, it will throw an error.
4. If a task to be run on a existing cluster has a different `image_id` (even if it is the same in the same region the cluster is in), we will just error out.

Tested:
- [x] `sky launch -c test-image-dict test.yaml`
```
# test.yaml
resources:
  cloud: aws
  image_id:
    us-east-1: ami-0fb70a87623506e3e
    us-west-2: ami-03e40ee6f58eea284
```
- [x] `sky launch -c test-image-dict-failover test.yaml`
```
# test.yaml
resources:
  cloud: aws
  image_id:
    us-east-1: ami-0fb70a87623506e3e
    us-west-2: invalid
```
- [x] `tests/run_smoke_tests.sh test_image_id_dict`
- [x] `tests/run_smoke_tests.sh test_image_id_dict_with_region`
- [x] `tests/run_smoke_tests.sh test_image_id_dict_with_zone`

### CODE REVIEW

**Comment by concretevitamin** [comment]
May not be able to get to this today. Tagging @WoosukKwon as well.

**Comment by romilbhardwaj** [comment]
Taking a look at this now

File: sky/resources.py [hunk 1]
```diff
@@ -114,7 +116,10 @@ def __repr__(self) -> str:
 
         image_id = ''
         if self.image_id is not None:
-            image_id = f', image_id={self.image_id!r}'
+            if None in self.image_id:
+                image_id = f', image_id={self.image_id[None]}'
+            else:
+                image_id = f', image_id={self.image_id!r}'
```
Comment (romilbhardwaj) [inline sky/resources.py:122]: If there are many image ids, `sky status` becomes a little crowded:
```
NAME  LAUNCHED        RESOURCES                                                                  STATUS  AUTOSTOP  COMMAND
clus  a few secs ago  1x AWS(m6i.2xlarge, image_id={'us-east-1': 'ami-0fb70a87623506e3e', 'u...  UP      -         sky launch -c clus test.y...
```

I'm trying to think of a better way to do this but not sure what's the best:
1. List only regions/zones: `image_id={us-east-1,us-west-1}`. Pros - gives a hint which images are specified. Cons - can be confusing, no details about the specific image used.
2. Show the count of images specified: `image_id={5 images}`. Pros - concise. Cons - no details about region/images
3. Do not show image_id in sky status. Pros - concise. Cons - No details.
4. Keep as it is. Pros - detailed. Cons - too long

Comment (romilbhardwaj) [reply sky/resources.py:122]: On thinking more, this field is shown only if image_id is specified. And if a user is specifying an image_id, they would want to know which image is being used for a task. So choosing option 4 (keep as is) might be okay.

File: tests/test_smoke.py [hunk 2]
```diff
@@ -155,8 +155,9 @@ def test_region():
             f'sky launch -y -c {name} --region us-west-2 examples/minimal.yaml',
             f'sky exec {name} examples/minimal.yaml',
             f'sky logs {name} 1 --status',  # Ensure the job succeeded.
+            f'sky status --all | grep {name} | grep us-west-2',  # Ensure the region is correct.
         ],
-        f'sky down -y {name}',
+        f'sky down -y {name} {name}-2 {name}-3',
```
Comment (romilbhardwaj) [inline tests/test_smoke.py:160]: Is this change required? Not sure how `{name}-2` and `{name}-3` clusters are being created. Same at L176

File: tests/test_smoke.py [hunk 3]
```diff
@@ -172,6 +173,82 @@ def test_zone():
...
+
+
+def test_image_id_dict():
+    name = _get_cluster_name()
+    test = Test(
+        'image_id_dict',
+        [
+            # Use image id dict.
+            f'sky launch -y -c {name} examples/per_region_images.yaml',
+            f'sky exec {name} examples/per_region_images.py',
```
Comment (romilbhardwaj) [inline tests/test_smoke.py:188]: I think these two files (`per_region_images.py/yaml`) are not included in this PR :)

File: tests/test_smoke.py [hunk 4]
```diff
@@ -172,6 +173,82 @@ def test_zone():
...
+def test_image_id_dict_with_region():
+    name = _get_cluster_name()
+    test = Test(
+        'image_id_dict_with_region',
+        [
+            # Use region to filter image_id dict.
+            f'sky launch -y -c {name} --region us-west-2 examples/per_region_images.yaml && exit 1 || true',
+            f'sky status | grep {name} && exit 1 || true',  # Ensure the cluster is not created.
+            f'sky launch -y -c {name} --region us-west-1 examples/per_region_images.yaml',
+            # Fail the task if even if the image id match for the region (image_id should be the same dict).
```
Comment (romilbhardwaj) [inline tests/test_smoke.py:208]: nit: typo: "Fail the task ~if~ even if"

**Review by romilbhardwaj — COMMENTED** [review]
Thanks for the awesome work @Michaelvll! Left some comments and faced a bug, mentioned below.

**Comment by romilbhardwaj** [comment]
Bug:
I created a cluster named `clus` with:
```
resources:
  cloud: aws
  image_id:
    us-east-1: ami-0fb70a87623506e3e
    us-west-2: ami-03e40ee6f58eea284
```

My cluster got provisioned on us-east-1. Then I removed `us-west-2` from image_id dict:
```
resources:
  cloud: aws
  image_id:
    us-east-1: ami-0fb70a87623506e3e
```

However I get this error, even though I feel it should work:
```
$ sky launch -c clus test.yaml
Task from YAML spec: test.yaml
W 11-11 10:14:45 resources.py:638] image_id in resources is experimental. It only supports AWS/GCP.
sky.exceptions.ResourcesMismatchError: Requested resources do not match the existing cluster.
  Requested:    1x AWS(None, image_id={'us-east-1': 'ami-0fb70a87623506e3e'})
  Existing:     1x AWS(m6i.2xlarge, image_id={'us-east-1': 'ami-0fb70a87623506e3e', 'us-west-2': 'ami-03e40ee6f58eea284'})
To fix: specify a new cluster name, or down the existing cluster first: sky down clus
```

We might need to update the `less_demanding_than` method to check if the requested resources image_id dict is a subset of the existing resources image_id dict/str:

https://github.com/skypilot-org/skypilot/blob/40ab453f2faf14183ed03aaedf49a004aef84519/sky/resources.py#L519-L520

**Comment by Michaelvll** [comment]
Thanks for the review @romilbhardwaj! Previously, I intentionally requires the image dict in a task to be the same as launched resources, but after reading your comment and thinking more, I think it makes sense to allow the task to have a less demanding image_id than the cluster. I updated the code.
The current behavior is that:
The task `image_id` is not specified or contains the same image in the region (even if it has some different images request for other regions) the existing cluster is in: succeed.

**Comment by Michaelvll** [comment]
Tested:
- [ ] `tests/run_smoke_tests.sh`

File: sky/resources.py (hunk 1 above)
Comment (romilbhardwaj) [reply sky/resources.py:124]: ^that's an excellent idea! I like it

**Review by romilbhardwaj — COMMENTED** [review]
Nice! Once we have https://github.com/skypilot-org/skypilot/pull/1384#discussion_r1020580059 also added to this PR, we should be good to go!

**Comment by Michaelvll** [comment]
> Nice! Once we have [#1384 (comment)](https://github.com/skypilot-org/skypilot/pull/1384#discussion_r1020580059) also added to this PR, we should be good to go!

This should have been added. PTAL. : )

**Comment by romilbhardwaj** [comment]
> > Nice! Once we have [#1384 (comment)](https://github.com/skypilot-org/skypilot/pull/1384#discussion_r1020580059) also added to this PR, we should be good to go!
> 
> This should have been added. PTAL. : )

Oops - I hadn't updated my branch. Looks great to me!
```
NAME    LAUNCHED    RESOURCES                                                             STATUS  AUTOSTOP  COMMAND
image   1 min ago   1x AWS(m6i.2xlarge, image_id={'us-east-1': 'ami-0fb70a87623506e3e'})  UP      -         sky launch -c image test....
```

**Review by romilbhardwaj — APPROVED** [review]
lgtm! :)
